ocaml/xenopsd/xc/service.ml

```diff
@@ -256,31 +258,56 @@ module DaemonMgmt (D : DAEMONPIDPATH) = struct
 
   let name = D.name
 
-  let pid ~xs domid =
+  (* Look up the commandline args for the process pid; *)
+  (* Check that they include expected args. *)
+  let is_cmdline_valid pid expected_args =
     try
-      match D.pid_location with
-      | (File file | Path_of {file; _}) when Sys.file_exists (file domid) ->
-          let path = file domid in
-          let ( let* ) = Option.bind in
-          let* pid = Unixext.pidfile_read path in
-          Unixext.with_file path [Unix.O_RDONLY] 0 (fun fd ->
-              try
-                Unix.lockf fd Unix.F_TRLOCK 0 ;
-                (* we succeeded taking the lock: original process is dead.
-                 * some other process might've reused its pid *)
-                None
-              with Unix.Unix_error (Unix.EAGAIN, _, _) ->
-                (* cannot obtain lock: process is alive *)
-                Some pid
-          )
-      | Xenstore key | Path_of {key; _} ->
-          (* backward compatibility during update installation: only has
-             xenstore pid available *)
-          let pid = xs.Xs.read (key domid) in
-          Some (int_of_string pid)
-      | _ ->
+      let cmdline =
+        Printf.sprintf "/proc/%d/cmdline" pid
+        |> Unixext.string_of_file
+        |> Astring.String.cuts ~sep:"\000"
+      in
+      List.for_all (fun arg -> List.mem arg cmdline) expected_args
```

Comment:
might be useful to add a comment here that this relies on the cmdline containing the domid (which it currently does with `-domid ...`, but just in case we change that in the future).
Otherwise we'd still risk killing another qemu instead of the correct one.

Reply:
Comments updated: https://github.com/xapi-project/xen-api/pull/5325/commits/8b8d84df18368975f52167c42e8f810107542c00